extensions/amp-story/1.0/amp-story.js

```diff
@@ -1258,17 +1319,43 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
     if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      pageIndex >= PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.PENDING
     ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
-      }
-      this.storeService_.dispatch(Action.TOGGLE_SUBSCRIPTIONS_DIALOG, true);
+      return this.blockOnPendingSubscriptionsState_(targetPageId, direction);
+    }
 
-      // TODO(#37285): add SubscriptionService to actually trigger the subscription dialog.
+    // Navigation to the locked pages after the paywall page should be redirected to the paywall page.
+    if (
+      pageIndex > PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
+    ) {
+      this.pageAfterGranted_ = targetPageId;
+      this.showSubscriptionsDialog_();
+      return this.switchTo_(
```

Comment:
Is this `switchTo_` for deeplinking cases? Can you add an inline comment?

Reply:
At this point the navigation hasn't happened yet so you could just return, but that's ok. Can you add a comment to say it's necessary in case of deeplinking tho plz?